File: internal/updater/discovery.go
// Package updater discovers, verifies, and applies p-track releases.
package updater

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	latestReleaseURL = "https://api.github.com/repos/ro-ag/ptrack/releases/latest"
	userAgent        = "p-track-updater"
	maxMetadataBytes = 1 << 20
	maxNotesBytes    = 32 << 10
	maxManifestBytes = 1 << 20
	maxAssetBytes    = 512 << 20
)

var (
	ErrDevelopmentBuild  = errors.New("updates are unavailable for development builds")
	ErrInvalidRelease    = errors.New("invalid GitHub release")
	ErrNoUpdate          = errors.New("no newer release is available")
	ErrUnsupportedTarget = errors.New("unsupported update target")
)

// Target is a host-selected operating system and architecture pair. It must
// never be populated from frontend or release metadata.
type Target struct {
	GOOS   string
	GOARCH string
}

// Asset is one exact packaged GitHub Release asset selected by the host.
type Asset struct {
	Name        string
	DownloadURL string
	SizeBytes   int64
}

// Candidate is a newer stable release and the two assets required to stage it.
type Candidate struct {
	Version     string
	Tag         string
	PageURL     string
	PublishedAt time.Time
	Notes       string
	Package     Asset
	Checksums   Asset
}

// Client performs bounded discovery against p-track's fixed GitHub repository.
type Client struct {
	endpoint string
	client   *http.Client
}

// NewClient returns a production release-discovery client. Redirects are not
// accepted for metadata; release asset redirects are handled during staging.
func NewClient() *Client {
	return &Client{
		endpoint: latestReleaseURL,
		client: &http.Client{
			Timeout: 15 * time.Second,
			CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
				return http.ErrUseLastResponse
			},
		},
	}
}

// Check returns a strictly newer stable release for target. It accepts only
// p-track's exact packaged asset names and checksum manifest.
func (c *Client) Check(ctx context.Context, currentVersion string, target Target) (Candidate, error) {
	current, err := parseVersion(currentVersion, true)
	if err != nil {
		return Candidate{}, ErrDevelopmentBuild
	}
	wantedName, err := packageName(target, "VERSION")
	if err != nil {
		return Candidate{}, err
	}

	release, err := c.fetch(ctx)
	if err != nil {
		return Candidate{}, err
	}
	remote, err := parseVersion(release.TagName, false)
	if err != nil || release.TagName != "v"+remote.String() || release.Draft || release.Prerelease {
		return Candidate{}, fmt.Errorf("%w: release must be a published stable vX.Y.Z tag", ErrInvalidRelease)
	}
	if remote.compare(current) <= 0 {
		return Candidate{}, ErrNoUpdate
	}
	if release.PublishedAt.IsZero() || len(release.Body) > maxNotesBytes {
		return Candidate{}, fmt.Errorf("%w: missing publication time or oversized notes", ErrInvalidRelease)
	}

	wantedName = strings.Replace(wantedName, "VERSION", remote.String(), 1)
	packageAsset, err := selectAsset(release.Assets, release.TagName, wantedName, maxAssetBytes)
	if err != nil {
		return Candidate{}, err
	}
	checksums, err := selectAsset(release.Assets, release.TagName, "checksums.txt", maxManifestBytes)
	if err != nil {
		return Candidate{}, err
	}

	return Candidate{
		Version:     remote.String(),
		Tag:         release.TagName,
		PageURL:     "https://github.com/ro-ag/ptrack/releases/tag/" + release.TagName,
		PublishedAt: release.PublishedAt.UTC(),
		Notes:       release.Body,
		Package:     packageAsset,
		Checksums:   checksums,
	}, nil
}

type githubRelease struct {
	TagName     string        `json:"tag_name"`
	Body        string        `json:"body"`
	Draft       bool          `json:"draft"`
	Prerelease  bool          `json:"prerelease"`
	PublishedAt time.Time     `json:"published_at"`
	Assets      []githubAsset `json:"assets"`
}

type githubAsset struct {
	Name               string `json:"name"`
	BrowserDownloadURL string `json:"browser_download_url"`
	Size               int64  `json:"size"`
	State              string `json:"state"`
}

func (c *Client) fetch(ctx context.Context) (githubRelease, error) {
	if c == nil || c.client == nil || c.endpoint == "" {
		return githubRelease{}, fmt.Errorf("%w: release client is not configured", ErrInvalidRelease)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.endpoint, nil)
	if err != nil {
		return githubRelease{}, fmt.Errorf("build release request: %w", err)
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("X-GitHub-Api-Version", "2022-11-28")

	httpClient := *c.client
	httpClient.CheckRedirect = func(_ *http.Request, _ []*http.Request) error {
		return http.ErrUseLastResponse
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return githubRelease{}, fmt.Errorf("fetch GitHub release: %w", err)
	}
	defer resp.Body.Close()
	if resp.Request == nil || resp.Request.URL == nil || resp.Request.URL.String() != c.endpoint {
		return githubRelease{}, fmt.Errorf("%w: metadata redirect refused", ErrInvalidRelease)
	}
	if resp.StatusCode >= http.StatusMultipleChoices && resp.StatusCode < http.StatusBadRequest {
		return githubRelease{}, fmt.Errorf("%w: metadata redirect refused", ErrInvalidRelease)
	}
	if resp.StatusCode != http.StatusOK {
		return githubRelease{}, fmt.Errorf("fetch GitHub release: unexpected HTTP status %d", resp.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, maxMetadataBytes+1))
	if err != nil {
		return githubRelease{}, fmt.Errorf("read GitHub release: %w", err)
	}
	if len(body) > maxMetadataBytes {
		return githubRelease{}, fmt.Errorf("%w: metadata exceeds %d bytes", ErrInvalidRelease, maxMetadataBytes)
	}
	var release githubRelease
	if err := json.Unmarshal(body, &release); err != nil {
		return githubRelease{}, fmt.Errorf("%w: decode metadata: %v", ErrInvalidRelease, err)
	}
	return release, nil
}

func packageName(target Target, version string) (string, error) {
	if target.GOARCH != "amd64" && target.GOARCH != "arm64" {
		return "", fmt.Errorf("%w: %s/%s", ErrUnsupportedTarget, target.GOOS, target.GOARCH)
	}
	switch target.GOOS {
	case "darwin":
		return fmt.Sprintf("p-track_%s_darwin_%s.dmg", version, target.GOARCH), nil
	case "linux":
		return fmt.Sprintf("ptrack_%s_linux_%s.tar.gz", version, target.GOARCH), nil
	case "windows":
		return fmt.Sprintf("ptrack_%s_windows_%s.zip", version, target.GOARCH), nil
	default:
		return "", fmt.Errorf("%w: %s/%s", ErrUnsupportedTarget, target.GOOS, target.GOARCH)
	}
}

func selectAsset(assets []githubAsset, tag, name string, maxSize int64) (Asset, error) {
	var selected *githubAsset
	for i := range assets {
		if assets[i].Name != name {
			continue
		}
		if selected != nil {
			return Asset{}, fmt.Errorf("%w: duplicate asset %q", ErrInvalidRelease, name)
		}
		selected = &assets[i]
	}
	if selected == nil {
		return Asset{}, fmt.Errorf("%w: missing asset %q", ErrInvalidRelease, name)
	}
	if selected.State != "uploaded" || selected.Size <= 0 || selected.Size > maxSize {
		return Asset{}, fmt.Errorf("%w: invalid asset %q", ErrInvalidRelease, name)
	}
	if err := validateAssetURL(selected.BrowserDownloadURL, tag, name); err != nil {
		return Asset{}, err
	}
	return Asset{Name: name, DownloadURL: selected.BrowserDownloadURL, SizeBytes: selected.Size}, nil
}

func validateAssetURL(rawURL, tag, name string) error {
	u, err := url.Parse(rawURL)
	if err != nil || u.Scheme != "https" || u.Host != "github.com" || u.User != nil ||
		u.RawQuery != "" || u.ForceQuery || u.Fragment != "" || strings.ContainsAny(rawURL, "?#") {
		return fmt.Errorf("%w: unsafe URL for asset %q", ErrInvalidRelease, name)
	}
	wantPath := "/ro-ag/ptrack/releases/download/" + tag + "/" + name
	if u.EscapedPath() != wantPath || u.Path != wantPath {
		return fmt.Errorf("%w: unexpected URL for asset %q", ErrInvalidRelease, name)
	}
	return nil
}

type semVersion struct {
	major uint64
	minor uint64
	patch uint64
}

func parseVersion(raw string, allowOptionalV bool) (semVersion, error) {
	if allowOptionalV {
		raw = strings.TrimPrefix(raw, "v")
	} else if !strings.HasPrefix(raw, "v") {
		return semVersion{}, errors.New("version must start with v")
	} else {
		raw = strings.TrimPrefix(raw, "v")
	}
	parts := strings.Split(raw, ".")
	if len(parts) != 3 {
		return semVersion{}, errors.New("version must have three components")
	}
	values := make([]uint64, 3)
	for i, part := range parts {
		if part == "" || (len(part) > 1 && part[0] == '0') {
			return semVersion{}, errors.New("invalid numeric version component")
		}
		value, err := strconv.ParseUint(part, 10, 64)
		if err != nil {
			return semVersion{}, errors.New("invalid numeric version component")
		}
		values[i] = value
	}
	return semVersion{major: values[0], minor: values[1], patch: values[2]}, nil
}

func (v semVersion) String() string {
	return fmt.Sprintf("%d.%d.%d", v.major, v.minor, v.patch)
}

func (v semVersion) compare(other semVersion) int {
	left := [...]uint64{v.major, v.minor, v.patch}
	right := [...]uint64{other.major, other.minor, other.patch}
	for i := range left {
		if left[i] < right[i] {
			return -1
		}
		if left[i] > right[i] {
			return 1
		}
	}
	return 0
}

// CompareVersions compares two strict release versions (with optional leading
// v) and returns -1, 0, or 1. Development and prerelease strings are rejected.
func CompareVersions(left, right string) (int, error) {
	leftVersion, err := parseVersion(left, true)
	if err != nil {
		return 0, err
	}
	rightVersion, err := parseVersion(right, true)
	if err != nil {
		return 0, err
	}
	return leftVersion.compare(rightVersion), nil
}

File: crates/ptrack-updater/src/discovery.rs
use std::fmt;
use std::time::Duration;

use futures_util::StreamExt;
use reqwest::header::{ACCEPT, USER_AGENT};
use serde::Deserialize;
use time::OffsetDateTime;
use time::format_description::well_known::Rfc3339;
use tokio_util::sync::CancellationToken;

pub const LATEST_RELEASE_URL: &str = "https://api.github.com/repos/ro-ag/ptrack/releases/latest";
const USER_AGENT_VALUE: &str = "p-track-updater";
const MAX_METADATA_BYTES: usize = 1 << 20;
const MAX_NOTES_BYTES: usize = 32 << 10;
const MAX_MANIFEST_BYTES: u64 = 1 << 20;
const MAX_ASSET_BYTES: u64 = 512 << 20;

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum UpdateError {
    DevelopmentBuild,
    InvalidRelease,
    NoUpdate,
    UnsupportedTarget,
    InvalidStage,
    InstallRefused,
    PendingStageMismatch,
    Cancelled,
    Message(String),
}

impl fmt::Display for UpdateError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::DevelopmentBuild => "updates are unavailable for development builds",
            Self::InvalidRelease => "invalid GitHub release",
            Self::NoUpdate => "no newer release is available",
            Self::UnsupportedTarget => "unsupported update target",
            Self::InvalidStage => "invalid staged update",
            Self::InstallRefused => "update installation refused",
            Self::PendingStageMismatch => "pending update belongs to another verified stage",
            Self::Cancelled => "update operation was canceled",
            Self::Message(message) => message,
        })
    }
}

impl std::error::Error for UpdateError {}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Target {
    pub os: String,
    pub arch: String,
}

impl Target {
    #[must_use]
    pub fn host() -> Self {
        Self {
            os: match std::env::consts::OS {
                "macos" => "darwin",
                other => other,
            }
            .to_owned(),
            arch: match std::env::consts::ARCH {
                "x86_64" => "amd64",
                "aarch64" => "arm64",
                other => other,
            }
            .to_owned(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Asset {
    pub name: String,
    pub download_url: String,
    pub size_bytes: u64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Candidate {
    pub version: String,
    pub tag: String,
    pub page_url: String,
    pub published_at: String,
    pub notes: String,
    pub package: Asset,
    pub checksums: Asset,
}

pub struct Client {
    endpoint: String,
    http: reqwest::Client,
    #[cfg(test)]
    test_asset_base: Option<String>,
    #[cfg(test)]
    test_asset_http: Option<reqwest::Client>,
}

impl Client {
    /// Constructs the fixed, redirect-refusing production discovery client.
    ///
    /// # Errors
    /// Returns an error when the TLS or HTTP client cannot be configured.
    pub fn new() -> Result<Self, UpdateError> {
        let _ = rustls::crypto::ring::default_provider().install_default();
        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(15))
            .redirect(reqwest::redirect::Policy::none())
            .build()
            .map_err(|_| UpdateError::Message("configure release client".to_owned()))?;
        Ok(Self {
            endpoint: LATEST_RELEASE_URL.to_owned(),
            http,
            #[cfg(test)]
            test_asset_base: None,
            #[cfg(test)]
            test_asset_http: None,
        })
    }

    #[cfg(test)]
    pub(crate) fn with_endpoint(endpoint: String) -> Result<Self, UpdateError> {
        let mut value = Self::new()?;
        value.endpoint = endpoint;
        Ok(value)
    }

    #[cfg(test)]
    pub(crate) fn with_test_asset_server(base: String) -> Result<Self, UpdateError> {
        let mut value = Self::new()?;
        value.test_asset_base = Some(base);
        value.test_asset_http = Some(
            reqwest::Client::builder()
                .timeout(Duration::from_secs(30))
                .redirect(reqwest::redirect::Policy::none())
                .build()
                .map_err(|_| UpdateError::InvalidStage)?,
        );
        Ok(value)
    }

    #[cfg(test)]
    pub(crate) fn test_asset_transport(&self) -> Option<(&str, reqwest::Client)> {
        Some((
            self.test_asset_base.as_deref()?,
            self.test_asset_http.clone()?,
        ))
    }

    /// Selects the exact newer package and checksum assets for `target`.
    ///
    /// # Errors
    /// Returns a bounded discovery, cancellation, target, or release error.
    pub async fn check(
        &self,
        cancellation: &CancellationToken,
        current_version: &str,
        target: &Target,
    ) -> Result<Candidate, UpdateError> {
        let current =
            parse_version(current_version, true).map_err(|_| UpdateError::DevelopmentBuild)?;
        package_name(target, "VERSION")?;
        let request = self
            .http
            .get(&self.endpoint)
            .header(ACCEPT, "application/vnd.github+json")
            .header(USER_AGENT, USER_AGENT_VALUE)
            .header("X-GitHub-Api-Version", "2022-11-28")
            .build()
            .map_err(|_| UpdateError::InvalidRelease)?;
        let response = tokio::select! {
            () = cancellation.cancelled() => return Err(UpdateError::Cancelled),
            response = self.http.execute(request) => response.map_err(|_| UpdateError::InvalidRelease)?,
        };
        if response.url().as_str() != self.endpoint || response.status() != reqwest::StatusCode::OK
        {
            return Err(UpdateError::InvalidRelease);
        }
        let mut stream = response.bytes_stream();
        let mut bytes = Vec::new();
        loop {
            let next = tokio::select! {
                () = cancellation.cancelled() => return Err(UpdateError::Cancelled),
                next = stream.next() => next,
            };
            let Some(chunk) = next else { break };
            let chunk = chunk.map_err(|_| UpdateError::InvalidRelease)?;
            if chunk.len() > MAX_METADATA_BYTES.saturating_sub(bytes.len()) {
                return Err(UpdateError::InvalidRelease);
            }
            bytes.extend_from_slice(&chunk);
        }
        select_candidate(bytes.as_slice(), current, target)
    }
}

#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct Version([u64; 3]);

impl fmt::Display for Version {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}.{}.{}", self.0[0], self.0[1], self.0[2])
    }
}

/// Parses only stable three-component versions.
///
/// # Errors
/// Rejects whitespace, prerelease/build data, leading zeros, and overflow.
pub fn parse_version(value: &str, allow_optional_v: bool) -> Result<Version, UpdateError> {
    let value = if allow_optional_v {
        value.strip_prefix('v').unwrap_or(value)
    } else {
        value.strip_prefix('v').ok_or(UpdateError::InvalidRelease)?
    };
    if value.is_empty() || value.trim() != value {
        return Err(UpdateError::InvalidRelease);
    }
    let parts = value.split('.').collect::<Vec<_>>();
    if parts.len() != 3
        || parts.iter().any(|part| {
            part.is_empty()
                || !part.bytes().all(|byte| byte.is_ascii_digit())
                || (part.len() > 1 && part.starts_with('0'))
        })
    {
        return Err(UpdateError::InvalidRelease);
    }
    let numbers = parts
        .iter()
        .map(|part| part.parse::<u64>().map_err(|_| UpdateError::InvalidRelease))
        .collect::<Result<Vec<_>, _>>()?;
    Ok(Version([numbers[0], numbers[1], numbers[2]]))
}

/// Compares two strict stable versions.
///
/// # Errors
/// Returns an error when either input is not strict stable `X.Y.Z`.
pub fn compare_versions(left: &str, right: &str) -> Result<std::cmp::Ordering, UpdateError> {
    Ok(parse_version(left, true)?.cmp(&parse_version(right, true)?))
}

/// Returns the exact packaged release filename for one supported target.
///
/// # Errors
/// Rejects every target outside darwin/linux/windows × amd64/arm64.
pub fn package_name(target: &Target, version: &str) -> Result<String, UpdateError> {
    if !matches!(target.arch.as_str(), "amd64" | "arm64") {
        return Err(UpdateError::UnsupportedTarget);
    }
    match target.os.as_str() {
        "darwin" => Ok(format!("p-track_{version}_darwin_{}.dmg", target.arch)),
        "linux" => Ok(format!("ptrack_{version}_linux_{}.tar.gz", target.arch)),
        "windows" => Ok(format!("ptrack_{version}_windows_{}.zip", target.arch)),
        _ => Err(UpdateError::UnsupportedTarget),
    }
}

#[derive(Deserialize)]
struct GithubRelease {
    tag_name: String,
    #[serde(default)]
    body: String,
    draft: bool,
    prerelease: bool,
    published_at: String,
    assets: Vec<GithubAsset>,
}

#[derive(Deserialize)]
struct GithubAsset {
    name: String,
    browser_download_url: String,
    size: i64,
    state: String,
}

pub(crate) fn select_candidate(
    bytes: &[u8],
    current: Version,
    target: &Target,
) -> Result<Candidate, UpdateError> {
    let release: GithubRelease =
        serde_json::from_slice(bytes).map_err(|_| UpdateError::InvalidRelease)?;
    let remote = parse_version(&release.tag_name, false)?;
    if release.tag_name != format!("v{remote}") || release.draft || release.prerelease {
        return Err(UpdateError::InvalidRelease);
    }
    if remote <= current {
        return Err(UpdateError::NoUpdate);
    }
    if release.notes_too_large() {
        return Err(UpdateError::InvalidRelease);
    }
    let published = OffsetDateTime::parse(&release.published_at, &Rfc3339)
        .map_err(|_| UpdateError::InvalidRelease)?;
    if published.unix_timestamp_nanos() == -62_135_596_800_000_000_000 {
        return Err(UpdateError::InvalidRelease);
    }
    let published_at = published
        .to_offset(time::UtcOffset::UTC)
        .format(&Rfc3339)
        .map_err(|_| UpdateError::InvalidRelease)?;
    let name = package_name(target, &remote.to_string())?;
    let package = select_asset(&release.assets, &release.tag_name, &name, MAX_ASSET_BYTES)?;
    let checksums = select_asset(
        &release.assets,
        &release.tag_name,
        "checksums.txt",
        MAX_MANIFEST_BYTES,
    )?;
    Ok(Candidate {
        version: remote.to_string(),
        tag: release.tag_name.clone(),
        page_url: format!(
            "https://github.com/ro-ag/ptrack/releases/tag/{}",
            release.tag_name
        ),
        published_at,
        notes: release.body,
        package,
        checksums,
    })
}

impl GithubRelease {
    fn notes_too_large(&self) -> bool {
        self.body.len() > MAX_NOTES_BYTES || self.published_at.is_empty()
    }
}

fn select_asset(
    assets: &[GithubAsset],
    tag: &str,
    name: &str,
    maximum: u64,
) -> Result<Asset, UpdateError> {
    let matches = assets
        .iter()
        .filter(|asset| asset.name == name)
        .collect::<Vec<_>>();
    let [asset] = matches.as_slice() else {
        return Err(UpdateError::InvalidRelease);
    };
    let size = u64::try_from(asset.size).map_err(|_| UpdateError::InvalidRelease)?;
    if asset.state != "uploaded" || size == 0 || size > maximum {
        return Err(UpdateError::InvalidRelease);
    }
    validate_asset_url(&asset.browser_download_url, tag, name)?;
    Ok(Asset {
        name: name.to_owned(),
        download_url: asset.browser_download_url.clone(),
        size_bytes: size,
    })
}

pub(crate) fn validate_asset_url(url: &str, tag: &str, name: &str) -> Result<(), UpdateError> {
    let expected = format!("https://github.com/ro-ag/ptrack/releases/download/{tag}/{name}");
    (url == expected)
        .then_some(())
        .ok_or(UpdateError::InvalidRelease)
}
